Adding BOARD_SIZE define in config.h

Alderi-Tokori opened the pull request:
Allows compiling for any reasonable odd board size (not tested with size = 1)

Alderi-Tokori commented:
This is a basic change adding a define into config.h allowing users to compile Leelaz with various board sizes. Due to the way winograd tiles are handled, I can only guarantee for it to be working on odd sizes (3x3, 5x5, 7x7, 9x9, etc.).

sethtroisi commented on src/OpenCL.cpp:
```diff
@@ -187,8 +187,8 @@ void __in_transform_eq(float x[4][4], __global float * restrict V, int offset, i
                            const int Ppad) {
-    const int W = 19;
-    const int H = 19;
+    const int W = BOARD_SIZE;
```
consider adding BOARD_SQUARES = BOARD_SIZE*BOARD_SIZE;
which can be substituted in dozens of places here.

sethtroisi commented on src/Training.cpp:
```diff
@@ -234,15 +234,15 @@ void Training::dump_training(int winner_color, OutputChunker& outchunk) {
                 out << std::hex << hexbyte;
             }
-            // 361 % 4 = 1 so the last bit goes by itself
+            // BOARD_SIZE * BOARD_SIZE % 4 = 1 so the last bit goes by itself
```
This is only true for odd sized boards.

sethtroisi commented on src/config.h:
```diff
@@ -29,6 +29,11 @@
+ * BOARD_SIZE: Define size of the board to compile Leela with, must be an odd number due to winograd tiles
+ */
+#define BOARD_SIZE 19
+
```
#if (BOARD_SIZE % 2 == 0)
#error code assumes odd board size remove at your own risk
#endif

Alderi-Tokori replied:
Nice, thank you for reviewing the code

jkiliani commented:
Is it actually necessary to use a define I config.h here? What about coding it so that board size is automatically read from the weight file?

gcp commented:
There is no ambiguity (the filter count in the policy head should identify it), but parts of the code size some arrays statically at compile time based on the max board size, so unless you want to compile in a huge value by default (wasting memory and slowing things down), much more changes are needed to support dynamic board sizes.

gcp commented on src/FastBoard.h:
```diff
@@ -40,7 +40,7 @@ class FastBoard {
         largest board supported
     */
-    static constexpr int MAXBOARDSIZE = 19;
+    static constexpr int MAXBOARDSIZE = BOARD_SIZE;
```
This no longer has any reason of existence.

gcp replied:
Maybe defining something here or elsewhere that is BOARD_SIZE * BOARD_SIZE (SQUARE_COUNT?) or so would save a lot of redundancy.

Square count sounds a bit odd because go is played on the intersections, but I can't think of anything better right now.

Alderi-Tokori replied:
Do you mean I can simply remove the line, or should I keep it with 19?

Alderi-Tokori replied:
As for the constant for BOARD_SIZE * BOARD_SIZE, maybe BOARD_SQUARES as @sethtroisi proposed, or something like INTERSECTION_COUNT?

gcp replied:
Let's take BOARD_SQUARES. The line above can go as MAXBOARDSIZE==BOARD_SIZE.

gcp commented on src/Im2Col.h:
```diff
@@ -70,7 +70,7 @@ template <>
                const std::vector<net_t>& input,
                std::vector<float>& output) {
-    constexpr unsigned int boardsize = 19;
+    constexpr unsigned int boardsize = BOARD_SIZE;
```
Same. Pointless now.

Alderi-Tokori replied:
I don't understand this, the variable is used and the function itself is called somewhere in the code, so it doesn't seem useless to me, but if I put back 19, it still works (despite the following assert that should stop the program). 

Should I remove the 3 lines ?
```CPP
    constexpr unsigned int boardsize = 19;
    auto outSize = size_t{channels * boardsize * boardsize};
    assert(output.size() == outSize);
```

gcp replied:
I mean that it's setting boardsize==BOARD_SIZE.

Just use BOARD_SIZE.

gcp commented on src/SGFTree.cpp:
```diff
@@ -123,7 +123,7 @@ void SGFTree::populate_states(void) {
             if (!m_properties.count("SZ")) {
                 // No size, but SGF spec defines default size for Go
-                m_properties.insert(std::make_pair("SZ", "19"));
+                m_properties.insert(std::make_pair("SZ", "BOARD_SIZE"));
```
This is wrong, see comment right above the change.

gcp commented on src/SGFTree.cpp:
```diff
@@ -153,8 +153,8 @@ void SGFTree::populate_states(void) {
-        // last ditch effort: if no GM or SZ, assume 19x19 Go here
-        int bsize = 19;
+        // last ditch effort: if no GM or SZ, assume BOARD_SIZExBOARD_SIZE Go here
+        int bsize = BOARD_SIZE;
```
Wrong for same reason as above.

gcp commented on src/Network.cpp:
```diff
@@ -643,7 +643,7 @@ void convolve(size_t outputs,
     // Weight shape (output, input, filter_size, filter_size)
     // 96 22 3 3
-    // outputs[96,19x19] = weights[96,22x3x3] x col[22x3x3,19x19]
+    // outputs[96,BOARD_SIZExBOARD_SIZE] = weights[96,22x3x3] x col[22x3x3,BOARD_SIZExBOARD_SIZE]
```
This was a an example with specific values, so this doesn't make a lot of sense now. (Also Zero has 18, not 22 inputs...)

Alderi-Tokori replied:
Same question, should I just remove the comment then or simply put back 19 in it?

gcp replied:
I'd put the 19x19 back in.

gcp commented on src/SGFTree.cpp:
```diff
@@ -290,7 +290,7 @@ int SGFTree::string_to_vertex(const std::string& movestring) const {
     }
 
-    if (m_state.board.get_boardsize() <= 19) {
+    if (m_state.board.get_boardsize() <= BOARD_SIZE) {
```
Wrong.

Alderi-Tokori commented:
Alright, I'm going to work on the requested changes this evening.

gcp commented:
Dirichlet noise should also be rescaled according to board size (see AZ paper).

Alderi-Tokori commented:
Good catch, it never occurred to me.

Alderi-Tokori commented:
I think a crude linear proportionality to the number of intersections might do it (after all, I've trained mine with the Dirichlet noise intended for 19x19 and I haven't had any major problem).

gcp commented:
I think the AZ paper set out their reasoning?

" Dirichlet noise Dir(α) was added to the prior probabilities in the
root node; this was scaled in inverse proportion to the approximate number of legal moves in a
typical position, to a value of α = {0.3, 0.15, 0.03} for chess, shogi and Go respectively."

Chess ~35 and 19x19 Go ~361

Alderi-Tokori commented:
Yes, but I'm not sure the average number of legal moves is actually linearly proportional to the number of intersections in go (might be, might not be), and I can't really have a look-up table in the program to have a perfect Dirichlet noise for each board size.

Alderi-Tokori commented:
There we go, everything should be in order.

l1t1 commented:
nice job

sethtroisi commented on src/FastBoard.cpp:
```diff
@@ -50,8 +50,8 @@ int FastBoard::get_boardsize(void) const {
-    assert(y >= 0 && y < MAXBOARDSIZE);
+    assert(x >= 0 && x < BOARD_SIZE);
+    assert(y >= 0 && y < BOARD_SIZE);
     assert(x >= 0 && x < m_boardsize);
```
long term we can remove m_boardsize.

Alderi-Tokori replied:
That'd mean removing get_boardsize() and replacing every call with BOARD_SIZE. I'll do this tonight.

sethtroisi commented on src/Im2Col.h:
```diff
@@ -27,8 +27,8 @@ template <unsigned long filter_size>
-    constexpr unsigned int width = 19;
+    constexpr unsigned int height = BOARD_SIZE;
+    constexpr unsigned int width = BOARD_SIZE;
     constexpr unsigned int channel_size = height * width;
```
BOARD_SQUARES

sethtroisi commented on src/OpenCL.cpp:
```diff
@@ -49,7 +49,7 @@ static std::string sourceCode_config = R"(
     #define vload_net_t(offset,p) ((p)[(offset)])
     #define vstore_net_t(data,offset,p) (((p)[(offset)])=(data))
-)";
+    #define BOARD_SIZE )" + std::to_string(BOARD_SIZE);
```
This doesn't look right

Alderi-Tokori replied:
It's the neatest way I've found in order to pass the BOARD_SIZE define into the OpenCl code source (the code doesn't have access to external defines). My aim with this is so the user only has to modify the board size in one place (namely config.h).

I'm ready to change this to some other thing if you have a better idea, but I don't have any other idea myself if I want to keep this "only has to modify the size in config.h" constraint.

sethtroisi commented on src/OpenCL.cpp:
```diff
@@ -127,14 +127,14 @@ __kernel void merge(
                         __global net_t * restrict out,
                         __private const int channels) {
-        // cl::NDRange global(outputs, 19*19);
+        // cl::NDRange global(outputs, BOARD_SIZE*BOARD_SIZE);
```
BOARD_SQUARES

Alderi-Tokori replied:
In OpenCL, I haven't defined BOARD_SQUARES : it seems not that useful to me given the only place it can be used is in this one commented line (Only place in OpenCL where there's a BOARD_SIZE*BOARD_SIZE). 

Do you really want me to replace this by BOARD_SQUARE?

sethtroisi commented on src/OpenCL.cpp:
```diff
@@ -127,14 +127,14 @@ __kernel void merge(
-        const int height = 19;
+        const int width = BOARD_SIZE;
+        const int height = BOARD_SIZE;
         const int boardsize = width * height;
```
BOARD_SQUARES

sethtroisi commented on src/SGFTree.cpp:
```diff
@@ -136,7 +136,7 @@ void SGFTree::populate_states(void) {
+        if (bsize <= BOARD_SIZE) {
             // Assume 7.5 komi if not specified
             m_state.init_game(bsize, 7.5f);
             valid_size = true;
```
line 157 has a 19 on it that should be replaced

Alderi-Tokori replied:
I'd think so, yes.

gcp replied:
> I don't think a 19x19 net would handle correctly a 9x9 board.

Indeed it would not due to the policy heads etc.

>if we haven't found either GM or SZ, it makes sense as a last effort to assume it's the right size

No, you should assume it's 19 because that is what the SGF spec says. If that's not BOARD_SIZE, give up.

>So in that case should we reject it if not specified and BOARD_SIZE != 19 to be consistent and follow SGF spec?

Exactly.

Alderi-Tokori commented:
@sethtroisi There you go, removing m_boardsize is quite a big change so I'm waiting for a review from @gcp before doing anything else.

sethtroisi commented:
@Alderi-Tokori Agreed, m_boardsize is quite a large change and I totally agree with holding off. I'd suggest doing it in a 2nd pull request because it will touch parameters and a dozen files which would certainly clutter this PR

Alderi-Tokori commented:
@sethtroisi Well, I've already done it in this one ^^' If it is rejected, I could always undo it though.

sethtroisi commented on src/GameState.cpp:
```diff
@@ -39,7 +39,7 @@ void GameState::init_game(int size, float komi) {
     game_history.emplace_back(std::make_shared<KoState>(*this));
 
-    m_timecontrol.set_boardsize(board.get_boardsize());
+    m_timecontrol.set_boardsize(BOARD_SIZE);
```
My bet is that this should be hardcoded in m_timecontrol now

Alderi-Tokori replied:
Hmm true, I'll do that if gcp validates the removing of m_boardsize.

gcp commented on src/Network.cpp:
```diff
@@ -749,7 +749,7 @@ void Network::forward_cpu(std::vector<float>& input,
 
     winograd_convolve3(output_channels, input, conv_weights[0], V, M, conv_out);
-    batchnorm<361>(output_channels, conv_out,
+    batchnorm<BOARD_SQUARES>(output_channels, conv_out,
```
Fix the alignment of the line below.

gcp commented on src/Network.cpp:
```diff
@@ -762,15 +762,15 @@ void Network::forward_cpu(std::vector<float>& input,
         winograd_convolve3(output_channels, conv_in,
 	                       conv_weights[i], V, M, conv_out);
-        batchnorm<361>(output_channels, conv_out,
+        batchnorm<BOARD_SQUARES>(output_channels, conv_out,
```
Same.

gcp commented on src/Network.cpp:
```diff
@@ -762,15 +762,15 @@ void Network::forward_cpu(std::vector<float>& input,
         winograd_convolve3(output_channels, conv_in,
 			               conv_weights[i + 1], V, M, conv_out);
-        batchnorm<361>(output_channels, conv_out,
+        batchnorm<BOARD_SQUARES>(output_channels, conv_out,
```
etc

gcp commented on src/OpenCL.cpp:
```diff
@@ -750,11 +750,11 @@ void OpenCL_Network::convolve1(int channels, int outputs,
-    // fixed for 19x19
-    constexpr int width = 19;
-    constexpr int height = 19;
+    // fixed for BOARD_SIZExBOARD_SIZE
```
It's no longer fixed now is it? :-)

Alderi-Tokori replied:
Ah yeah, I'll modify the comment. By the wawy, should I keep these variables in those functions or should I remove them and directly use BOARD_SIZE/BOARD_SQUARES in the code?

gcp replied:
Keep the variables for now.

gcp commented on src/SGFTree.cpp:
```diff
@@ -290,13 +291,13 @@ int SGFTree::string_to_vertex(const std::string& movestring) const {
     }
 
-    int bsize = m_state.board.get_boardsize();
+    int bsize = BOARD_SIZE;
```
This makes no sense. This code was checking whether we correctly found a usable board size at any point.

Alderi-Tokori replied:
The removing of m_boardsize (and the corresponding getter, was done after one of seth's comments, suggesting it wasn't necessary anymore.  Should I put it back in?

Alderi-Tokori replied:
(Or should I just remove this test, since board sizes != BOARD_SIZE will throw an exception in populate_state now)

Alderi-Tokori commented:
@gcp I'm assuming you're validating removing m_boardsize. I'll do another commit hardcoding BOARD_SIZE in TimeControl (as @sethtroisi suggested) next if you think this is a good idea?

gcp commented:
Yes, prefer separate pull requests.

Alderi-Tokori commented:
Alright, that's what I was thinking. I'm going to revert the changes and clean that up.

gcp commented:
Sorry about changing my mind about half of this code, but I was so happy about seeing the hardcoded 19's go that it clouded my judgement.

Alderi-Tokori commented:
Haha no problem, I'm not particularly in a hurry to see this PR merged, and when I committed these changes without having consulted you beforehand, I was already prepared to undo them.

Alderi-Tokori commented:
@gcp Ok, I've reverted the last two commits then nitpicked the other changes (aside from removing m_boardsize) from those reverted commits and added a few modifications.

gcp commented on src/Im2Col.h:
```diff
@@ -70,8 +69,7 @@ template <>
                std::vector<float>& output) {
-    constexpr unsigned int boardsize = 19;
-    auto outSize = size_t{channels * boardsize * boardsize};
+    auto outSize = size_t{channels * static_cast<size_t>(BOARD_SQUARES)};
```
Does this actually need both size_t casts?

Alderi-Tokori replied:
If I don't cast the value, it raises a warning which makes it fail the AppVeyor check.

gcp commented on src/SGFTree.cpp:
```diff
@@ -124,7 +124,9 @@ void SGFTree::populate_states(void) {
                 m_properties.insert(std::make_pair("SZ", "19"));
-                valid_size = true;
+                if (19 == BOARD_SIZE) {
+                    valid_size = true;
```
valid_size only says whether m_boardsize has been set, not whether we can deal with it or not. So I think you don't need changes here. Later on, you use valid_size to see what the board has been set at, and there's a check that it was either set or defaulted to BOARD_SIZE.

gcp commented:
Looks great, I had one last round of nits.

Alderi-Tokori commented:
@gcp Aside from the casting (which I don't think I can remove), everything should be done.
